src/java.base/share/classes/jdk/internal/classfile/CodeBuilder.java

```diff
@@ -0,0 +1,1385 @@
+        extends ClassfileBuilder<CodeElement, CodeBuilder>
+        permits CodeBuilder.BlockCodeBuilder, ChainedCodeBuilder, TerminalCodeBuilder, NonterminalCodeBuilder {
+
+    /**
+     * {@return the {@link CodeModel} representing the method body being transformed,
+     * if this code builder represents the transformation of some {@link CodeModel}}
+     */
+    Optional<CodeModel> original();
+
+    /** {@return a fresh unbound label} */
+    Label newLabel();
+
+    /** {@return the label associated with the beginning of the current block}
+     * If the current {@linkplain CodeBuilder} is not a "block" builder, such as
+     * those provided by {@link #block(Consumer)} or {@link #ifThenElse(Consumer, Consumer)},
+     * the current block will be the entire method body. */
+    Label startLabel();
+
+    /** {@return the label associated with the end of the current block}
+     * If the current {@linkplain CodeBuilder} is not a "block" builder, such as
+     * those provided by {@link #block(Consumer)} or {@link #ifThenElse(Consumer, Consumer)},
+     * the current block will be the entire method body. */
+    Label endLabel();
+
+    /**
+     * {@return the local variable slot associated with the receiver}.
+     *
+     * @throws IllegalStateException if this is not a static method
+     */
+    int receiverSlot();
+
+    /**
+     * {@return the local variable slot associated with the specified parameter}.
+     * The returned value is adjusted for the receiver slot (if the method is
+     * an instance method) and for the requirement that {@code long} and {@code double}
+     * values require two slots.
+     *
+     * @param paramNo the index of the parameter
+     */
+    int parameterSlot(int paramNo);
+
+    /**
+     * {@return the local variable slot of a fresh local variable}  This method
+     * makes reasonable efforts to determine which slots are in use and which
+     * are not.  When transforming a method, fresh locals begin at the {@code maxLocals}
+     * of the original method.  For a method being built directly, fresh locals
+     * begin after the last parameter slot.
+     *
+     * <p>If the current code builder is a "block" code builder provided by
+     * {@link #block(Consumer)}, {@link #ifThen(Consumer)}, or
+     * {@link #ifThenElse(Consumer, Consumer)}, at the end of the block, locals
+     * are reset to their value at the beginning of the block.
+     *
+     * @param typeKind the type of the local variable
+     */
+    int allocateLocal(TypeKind typeKind);
+
+    /**
+     * Builds code fragment generated by the handler and synchronously transformed.
+     *
+     * @param transform the transform to apply to the code generated by the handler
+     * @param handler handler that receives a {@linkplain CodeBuilder} to
+     * generate the code.
+     * @return this builder
+     */
+    default CodeBuilder transforming(CodeTransform transform, Consumer<CodeBuilder> handler) {
```

Comment:
The functionality of this method, `transforming`, and `ClassfileBuilder::transform`, are in effect equivalent in their transforming: adding the results of transformed code to the builder. They differ in the source of code elements.

The latter's behaviour can be implemented using the former, with a  consumer that passes all elements of a code model to the builder e.g. `builder -> model.forEach(builder::with)`.

The difference in naming initially confused me. To me this suggests the method names should be the same? (perhaps with the transformer being consistently the last argument?).

Reply:
The `CodeBuilder::transforming` solves a bit different use cases than all the other transform.
It is designed to be able to use code transformations on a code building handler within a single pass.
Main reason is support of features like `StackTracker` in a form of code transformation. `StackTracker` (or any other similar tool requiring to monitor or affect code building) is passed as a transformation of a code fragment, while it can immediately serve as a source of information necessary to generate follow-up bytecode of the same method (in the same pass). 
Example of such use case is here: 
https://github.com/openjdk/jdk/blob/0e43af667ba6c6bda61461c260688bc46d3f3474/src/java.base/share/classes/jdk/internal/classfile/components/CodeStackTracker.java#L49

These code generation/transformation cases must be handled in a single pass and `CodeBuilder::transforming` method has no similar peer in any other (method, field or class) builder, because it is not necessary.

Reply:
The use-case seems fine to me (and that it only makes sense for building code). I still think it's a "transform", but with a different source. Subtly changing the name makes it seem different and fundamentally it is not AFAICT. If there is a separate name I think it should reflect the difference in source input to the transformation, rather than differentiate via the present participle.

Reply:
I see what you mean. `transforming` was selected to represent the continuation of the code building process, similar to `trying` and `catching`. While `transform` may imply that the actual code builder gets transformed into something else. For example `MethodModel::transformCode` takes `CodeModel`, applies `CodeTransform` and after that the code of the actual method is finished.
What about `transformBlock(BlockCodeBuilder, CodeTransform)`?

Reply:
I am unsure how you might use `BlockCodeBuilder`. If the current signature is not changed then `transformFromHandler` seems reasonable (since its the handler that pushes elements into its given builder).

The other `transform` is implicitly "transform model".

Reply:
This topic is now tracked as RFE